Security/Principal/WindowsLoggedInIdentity.cs

```diff
@@ -36,23 +37,36 @@ public class WindowsLoggedInIdentity : IDisposable, IIdentity
 		/// The logon provider. This parameter can usually be left as the default. For more information, lookup more detail for the
 		/// dwLogonProvider parameter of the Windows LogonUser function.
 		/// </param>
-		public WindowsLoggedInIdentity(string userName, string domainName, string password, LogonUserType logonType = LogonUserType.LOGON32_LOGON_INTERACTIVE,
-			LogonUserProvider provider = LogonUserProvider.LOGON32_PROVIDER_DEFAULT)
+		public WindowsLoggedInIdentity(string userName, string domainName, string password, LogonUserType logonType = LogonUserType.LOGON32_LOGON_INTERACTIVE, LogonUserProvider provider = LogonUserProvider.LOGON32_PROVIDER_DEFAULT)
 		{
-			if (string.IsNullOrEmpty(userName)) throw new ArgumentNullException(nameof(userName));
-			if (string.IsNullOrEmpty(password)) throw new ArgumentNullException(nameof(password));
-			if (string.IsNullOrEmpty(domainName) && !userName.Contains("@")) throw new ArgumentNullException(nameof(domainName));
-			if (LogonUser(userName, domainName, password, logonType, provider, out hToken))
+			if (Extensions.StringHelper.IsNullOrWhiteSpace(userName))
+				throw new ArgumentNullException(nameof(userName));
+
+			if (Extensions.StringHelper.IsNullOrWhiteSpace(password))
+				throw new ArgumentNullException(nameof(password));
+
+			if (Extensions.StringHelper.IsNullOrWhiteSpace(domainName) && !userName.Contains("@"))
+				throw new ArgumentNullException(nameof(domainName));
+
+
+			var success = LogonUser(userName, domainName, password, logonType, provider, out hToken);
+
+			var lastError = Marshal.GetLastWin32Error();
+
+			if (success)
 			{
 				using (hToken)
 				{
 					AuthenticatedIdentity = new WindowsIdentity(hToken.DangerousGetHandle());
+					ownId = true;
```

Comment:
I can't see how this code changes the value of AuthenticatedIdentity. The values of ownId and AuthenticatedIdentitiy do not change any differently in this code until the disposal. Will you walk me through the problem and the logic you've used for fixing it?